iree/hal/vulkan/emulated_timeline_semaphore.cc

```diff
@@ -0,0 +1,461 @@
+  // Make sure the previous signaled value is smaller than the new value.
+  CHECK(signaled_value < value)
+      << "Attempting to signal a timeline value out of order; trying " << value
+      << " but " << signaled_value << " already signaled";
+
+  // Inform all queues to make progress given we have a new value signaled now.
+  for (CommandQueue* queue : logical_device_->queues()) {
+    RETURN_IF_ERROR(
+        static_cast<SerializingCommandQueue*>(queue)->AdvanceQueueSubmission());
+  }
+
+  return OkStatus();
+}
+
+Status EmulatedTimelineSemaphore::Wait(uint64_t value, absl::Time deadline) {
+  IREE_TRACE_SCOPE0("EmulatedTimelineSemaphore::Wait");
+
+  RETURN_IF_ERROR(TryToAdvanceTimeline(value));
+
+  // Collect the fences associated with all the time points that are before the
+  // deadline. We need to wait on them to be all signaled so that we can make
+  // sure the timeline is advanced to the required value.
+
+  absl::InlinedVector<VkFence, 4> fences;
+  {
+    absl::MutexLock lock(&mutex_);
+    for (auto* semaphore : outstanding_semaphores_) {
+      if (semaphore->value <= value) {
+        fences.push_back(semaphore->wait_fence->value());
+      } else {
+        break;
+      }
+    }
+  }
+
+  if (fences.empty()) return OkStatus();
+
+  uint64_t timeout_nanos;
+  if (deadline == absl::InfiniteFuture()) {
+    timeout_nanos = UINT64_MAX;
+  } else if (deadline == absl::InfinitePast()) {
+    timeout_nanos = 0;
+  } else {
+    auto relative_nanos = absl::ToInt64Nanoseconds(deadline - absl::Now());
+    timeout_nanos = relative_nanos < 0 ? 0 : relative_nanos;
+  }
+
+  VK_RETURN_IF_ERROR(logical_device_->syms()->vkWaitForFences(
+      *logical_device_, fences.size(), fences.data(), /*waitAll=*/true,
+      timeout_nanos));
+
+  RETURN_IF_ERROR(TryToAdvanceTimeline(value));
+  return OkStatus();
+}
+
+void EmulatedTimelineSemaphore::Fail(Status status) {
+  IREE_TRACE_SCOPE0("EmulatedTimelineSemaphore::Fail");
+  absl::MutexLock lock(&mutex_);
+  status_ = std::move(status);
+  signaled_value_.store(UINT64_MAX);
+}
+
+VkSemaphore EmulatedTimelineSemaphore::GetWaitSemaphore(
+    uint64_t value, const ref_ptr<TimePointFence>& wait_fence) {
+  IREE_TRACE_SCOPE0("EmulatedTimelineSemaphore::GetWaitSemaphore");
+  absl::MutexLock lock(&mutex_);
+
+  VkSemaphore semaphore = VK_NULL_HANDLE;
+  for (TimePointSemaphore* point : outstanding_semaphores_) {
+    if (point->value > value && point->wait_fence) {
+      point->wait_fence = add_ref(wait_fence);
+      semaphore = point->semaphore;
+      break;
+    }
+  }
+
+  return semaphore;
+}
+
+Status EmulatedTimelineSemaphore::CancelWaitSemaphore(VkSemaphore semaphore) {
+  IREE_TRACE_SCOPE0("EmulatedTimelineSemaphore::CancelWaitSemaphore");
+  absl::MutexLock lock(&mutex_);
+  for (TimePointSemaphore* point : outstanding_semaphores_) {
+    if (point->semaphore != semaphore) continue;
+
+    if (!point->wait_fence)
+      return InternalError("Time point wasn't waited before");
```

Comment:
you want to use the builder types for these as they capture source location information (InternalErrorBuilder(IREE_LOC) << "Time point " ...);